File: src/creational/abstract_factory/MazeGameTest.java
package creational.abstract_factory;

import creational.abstract_factory.base.Maze;
import creational.abstract_factory.base.MazeFactory;
import creational.abstract_factory.base.Side;
import creational.abstract_factory.bombed.BombedMazeFactory;
import creational.abstract_factory.bombed.BombedWall;
import creational.abstract_factory.bombed.RoomWithABomb;
import creational.abstract_factory.enchanted.DoorNeedingSpell;
import creational.abstract_factory.enchanted.EnchantedMazeFactory;
import creational.abstract_factory.enchanted.EnchantedRoom;
import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.assertTrue;

class MazeGameTest {
    @Test
    public void createNormalMaze() {
        Maze maze = MazeGame.createMaze(new MazeFactory());
        assertTrue(maze.getRoom(1) != null);
    }

    @Test
    public void createEnchantedMaze() {
        Maze maze = MazeGame.createMaze(new EnchantedMazeFactory());
        assertTrue(maze.getRoom(1) instanceof EnchantedRoom);
        assertTrue(maze.getRoom(1).getSide(Side.EAST) instanceof DoorNeedingSpell);
    }

    @Test
    public void createBombedMaze() {
        Maze maze = MazeGame.createMaze(new BombedMazeFactory());
        assertTrue(maze.getRoom(1) != null);
        assertTrue(maze.getRoom(1) instanceof RoomWithABomb);
        assertTrue(maze.getRoom(1).getSide(Side.NORTH) instanceof BombedWall);
    }
}

File: src/creational/abstract_factory_kotlin/MazeGameTest.kt
package creational.abstract_factory_kotlin

import base.maze.Side
import creational.abstract_factory_kotlin.bombed.BombedMazeFactory
import creational.abstract_factory_kotlin.bombed.BombedWall
import creational.abstract_factory_kotlin.bombed.RoomWithABomb
import creational.abstract_factory_kotlin.enchanted.DoorNeedingSpell
import creational.abstract_factory_kotlin.enchanted.EnchantedMazeFactory
import creational.abstract_factory_kotlin.enchanted.EnchantedRoom
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Test

internal class MazeGameTest {
    @Test
    fun createNormalMaze() {
        val maze = MazeGame.createMaze(MazeFactory())
        assertTrue(maze.getRoom(1) != null)
    }

    @Test
    fun createEnchantedMaze() {
        val maze = MazeGame.createMaze(EnchantedMazeFactory())
        assertTrue(maze.getRoom(1) is EnchantedRoom)
        assertTrue(maze.getRoom(1)?.getSide(Side.EAST) is DoorNeedingSpell)
    }

    @Test
    fun createBombedMaze() {
        val maze = MazeGame.createMaze(BombedMazeFactory())
        assertTrue(maze.getRoom(1) != null)
        assertTrue(maze.getRoom(1) is RoomWithABomb)
        assertTrue(maze.getRoom(1)?.getSide(Side.NORTH) is BombedWall)
    }
}